### DISCUSSION TITLE
[proxy] gdi->primary_buffer is empty

### DISCUSSION THREAD

**Discussion by gemone** [FreeRDP/FreeRDP#8068] [General]
proxy plugin can not load `gdi->primary_buffer`.

https://github.com/FreeRDP/FreeRDP/blob/dfa231c0a55b005af775b833f92f6bcd30363d77/server/proxy/modules/demo/demo.cpp#L180

test code:

```c
static BOOL demo_client_end_paint(proxyPlugin* plugin, proxyData* pdata, void* custom)
{
	pClientContext* pc = pdata->pc;

	WLog_INFO(TAG, "gdi data: %d", pc->context.gdi->primary_buffer[0], pc->context.gdi->primary_buffer[1]);

	WINPR_ASSERT(plugin);
	WINPR_ASSERT(pdata);
	WINPR_ASSERT(custom);

	WLog_INFO(TAG, "%s", __FUNCTION__);
	return TRUE;
}
```

and test:

1. `gdi->primary->hdc->hwnd->ninvalid`: 0;
2. `gdi->suppressOutput`: 0;
3. `gdi->bitmap_size`: 0;

target server: `windows server 2016`.

```ini
[Channels]
GFX=true
DisplayControl=true
Clipboard=true
AudioInput=true
AudioOutput=true
DeviceRedirection=true
VideoRedirection=true
CameraRedirection=true
RemoteApp=false
PassthroughIsBlacklist=true
Passthrough=
Intercept=

[GFXSettings]
DecodeGFX=true
```

Is this option not implemented yet? `DecodeGFX`

**Comment by mfleisz** [comment]
Not 100% sure on the status of `DecodeGFX` but from looking at your config you are passing through all channels. In order to be able to process GFX data in the proxy you will need to intercept the GFX channel

**Reply by gemone** [reply]
Hi, it should have nothing to do with `GFX`. 

Now there is no data for attributes in `GDI`.

**Comment by akallabeth** [comment]
The proxy does not work as a normal client does. (note that there are huge differences between `stable-2.0` and `master` version of proxy, I only discuss master as the `2.0` was a first implementation with various limitations)
So, to optimize performance the proxy does not decode the data of channels (or drawing operations) but passes them on.
If you want to do some recording/... you need to pass all the data to your own 'shadow' client using the same configuration as the real client and decode the data there to get images.
